raft/raft.go

```diff
@@ -611,6 +611,7 @@ func (r *raft) becomePreCandidate() {
 	// but doesn't change anything else. In particular it does not increase
 	// r.Term or change r.Vote.
 	r.step = stepCandidate
+	r.lead = None
```

Comment:
It is strange to change `r.lead` here when we do not change `r.Term`. `r.lead` is still the leader of that term. I think we should leave `r.lead` unchanged here, and if we change anything, we should change the `inLease` check at line 711. We can't become a pre-candidate if we've heard from the leader within the lease timeout, so if `state == StatePreCandidate` `inLease` must be false.

Reply:
@xiang90 I think thesis section 6.2 point out how to handle this situation. see https://github.com/coreos/etcd/pull/8334#discussion_r135178400